lib/pbench/server/filetree.py

```diff
@@ -771,63 +765,48 @@ def __getitem__(self, dataset_id: str) -> Tarball:
         except KeyError:
             raise TarballNotFound(dataset_id) from None
 
-    def _clean_empties(self, controller: str):
+    def _clean_empties(self):
         """
-        Remove empty controller directories from the RESULTS and INCOMING
-        trees. If there are no remaining tarballs in the ARCHIVE controller
-        directory, remove all empty state subdirectories; and, if the
-        controller directory is now empty remove it as well.
-
-        Args:
-            controller: Name of the controller to clean up
+        Remove all empty state sub-directories if there are no
+        remaining tarballs in the ARCHIVE directory.
         """
-        results = self.options.RESULTS / controller
-        self.delete_if_empty(results)
-        incoming = self.options.INCOMING / controller
-        self.delete_if_empty(incoming)
-        archive = self.options.ARCHIVE / controller
+        archive = self.options.ARCHIVE
         if archive.exists() and not any(archive.glob(f"*{Dataset.TARBALL_SUFFIX}")):
             for file in archive.iterdir():
-                if Controller.is_statedir(file):
+                if StateDIRS.is_statedir(file):
                     self.delete_if_empty(file)
-            self.delete_if_empty(archive)
-            del self.controllers[controller]
 
-    def _add_controller(self, directory: Path) -> None:
+    def _add_tarball(self, directory: Path) -> None:
         """
-        Create a new Controller object, add it to the set of known controllers,
-        and append the discovered datasets (tarballs) to the list of known
-        datasets (by internal dataset ID) and tarballs (by tarball base name).
+        Create a new StateDIRS object and append the discovered datasets
+        (tarballs) to the list of known datasets (by internal dataset ID)
+        and tarballs (by tarball base name).
 
         Args:
-            directory: A controller directory within the ARCHIVE tree
+            directory: the ARCHIVE tree
         """
-        controller = Controller(
+        state_dir = StateDIRS(
             directory, self.options.INCOMING, self.options.RESULTS, self.logger
         )
```

Comment:
Your new "state directories" object is really a singleton associated with the global `FileTree` rather than a hierarchy containing tarballs. I think the logic needs to be reworked more deeply to avoid the extra level. Let's just link tarballs to `FileTree` directly. It's not clear that we need the `StateDIRS` class at all, but it's certainly not duplicated and associated with subsets of tarballs like this.

Also, I don't understand the name change here to `_add_tarball`; you're creating a `StateDIRS`, which while it might implicitly discover a bunch of tarballs as the old `Controller` constructor did isn't "adding [a] tarball".